### PR TITLE
Updated PostgreSQL support

### PR DESCRIPTION
**Pull request by sergey-safarov**

### CODE REVIEW

File: kamailio/db_queries_postgres.cfg [hunk 1]
```diff
@@ -6,9 +6,11 @@
...
 #!substdef "!KZQ_PRESENCE_SEARCH_DETAIL!select * from active_watchers_log where presentity_uri = '\$var(presentity_uri)'!g"
 #!substdef "!KZQ_PRESENCE_SEARCH_SUMMARY!select * from active_watchers where watcher_domain = '\$var(Domain)'!g"
 #!substdef "!KZQ_PRESENCE_RESET!delete from presentity where sender = '\$var(MediaUrl)'!g"
+#!substdef "!KZQ_REPLACE_WATCHERS_LOG! DELETE FROM active_watchers_log WHERE presentity_uri='\$subs(uri)' and watcher_username='\$subs(watcher_username)' and watcher_domain='\$subs(watcher_domain)' and event='\$subs(event)'; INSERT INTO active_watchers_log (presentity_uri, watcher_username, watcher_domain, event, callid, to_user, to_domain, user_agent, time, result, sent_msg, received_msg) VALUES ('\$subs(uri)', '\$subs(watcher_username)', '\$subs(watcher_domain)', '\$subs(event)', '\$subs(callid)', '\$subs(to_user)', '\$subs(to_domain)', '\$(subs(user_agent){s.escape.common}{s.replace,',''}{s.replace,\$\$,})', \$TS, \$notify_reply(\$rs), '\$(mb{s.escape.common}{s.replace,',''}{s.replace,\$\$,}{s.replace,\\\\\",\"})', '\$(notify_reply(\$mb){s.escape.common}{s.replace,',''}{s.replace,\$\$,})')!g"
```
Comment (sergey-safarov) [inline kamailio/db_queries_postgres.cfg]: I added `{s.replace,\\\\\",\"}` that is removes `"` escaping in blob.
Is correct case when blob contains raw (not escaped) XML?

Comment (sergey-safarov) [reply kamailio/db_queries_postgres.cfg]: Also here in macro defined two SQL commands `DELETE` and `INSERT`.
This may be replaced by one `INSERT` with [`ON CONFLICT Clause`](https://www.postgresql.org/docs/9.5/static/sql-insert.html#SQL-ON-CONFLICT).
In this case must be used PostgreSQL 9.5 or newer.

**Comment by sergey-safarov** [comment]
@lazedo i updated macro
with small changes
1. Removed leading escape char.
2. preserved space symbols, this help parse so long string.

Current macro is:
```
#!substdef "!KZQ_REPLACE_WATCHERS_LOG!REPLACE INTO active_watchers_log (presentity_uri, watcher_username, watcher_domain, event, callid, to_user, to_domain, user_agent, time, result, sent_msg, received_msg) VALUES (\"\$subs(uri)\", \"\$subs(watcher_username)\", \"\$subs(watcher_domain)\", \"\$subs(event)\", \"\$subs(callid)\", \"\$subs(to_user)\", \"\$subs(to_domain)\", '\$(subs(user_agent){s.escape.common}{s.replace,\\\',''}{s.replace,\$\$,})', \$TS, \$notify_reply(\$rs), '\$(mb{s.escape.common}{s.replace,\\\',''}{s.replace,\$\$,})', '\$(notify_reply(\$mb){s.escape.common}{s.replace,\\\',''}{s.replace,\$\$,})')|g"
```
Technically i not be able test on your kamailio build with `db_kazoo` module. Now i use only reference kamailio repo with PostgreSQL database.
Same case with MySQL.
About kamailio with PostgreSQL i have tested suggested macro and see any errors.

**Comment by sergey-safarov** [comment]
Luis @lazedo I removed free space symbols.
Now `KZQ_REPLACE_WATCHERS_LOG` macro equal to your suggestions.

File: kamailio/db_queries_kazoo.cfg [hunk 2]
```diff
@@ -6,10 +6,11 @@
 #!substdef "!KZQ_EVENT_PRESENCE_RESET_DELETE!delete from presentity where domain=\"\$(kzE{kz.json,Realm})\" and username = \"\$(kzE{kz.json,Username})\"!g"
 #!substdef "!KZQ_HANDLE_NEW_SUBSCRIBE_DELETE1!delete from active_watchers where callid = \"\$ci\"!g"
 #!substdef "!KZQ_HANDLE_NEW_SUBSCRIBE_DELETE2!delete from active_watchers where watcher_username=\"\$fU\" and presentity_uri=\"\$var(presentity_uri)\" and to_user=\"\$tU\" and watcher_domain=\"\$fd\" and event=\"\$hdr(Event)\"!g"
+#!substdef "!KZQ_HAS_PRESENTITY!select count(*) as count from presentity where username = \"\$subs(to_user)\" and domain = \"\$subs(to_domain)\" and event = \"\$subs(event)\"!g"
```
Comment (lazedo) [inline kamailio/db_queries_kazoo.cfg]: why was this moved ?

File: kamailio/db_queries_mysql.cfg [hunk 3]
```diff
@@ -6,10 +6,11 @@
 #!substdef "!KZQ_EVENT_PRESENCE_RESET_DELETE!delete from presentity where domain=\"\$(kzE{kz.json,Realm})\" and username = \"\$(kzE{kz.json,Username})\"!g"
 #!substdef "!KZQ_HANDLE_NEW_SUBSCRIBE_DELETE1!delete from active_watchers where callid = \"\$ci\"!g"
 #!substdef "!KZQ_HANDLE_NEW_SUBSCRIBE_DELETE2!delete from active_watchers where watcher_username=\"\$fU\" and presentity_uri=\"\$var(presentity_uri)\" and to_user=\"\$tU\" and watcher_domain=\"\$fd\" and event=\"\$hdr(Event)\"!g"
+#!substdef "!KZQ_HAS_PRESENTITY!select count(*) as count from presentity where username = \"\$subs(to_user)\" and domain = \"\$subs(to_domain)\" and event = \"\$subs(event)\"!g"
```
Comment (lazedo) [inline kamailio/db_queries_mysql.cfg]: same here

File: kamailio/db_queries_postgres.cfg [hunk 4]
```diff
@@ -6,9 +6,11 @@
 #!substdef "!KZQ_EVENT_PRESENCE_RESET_DELETE!delete from presentity where domain='\$(kzE{kz.json,Realm})' and username = '\$(kzE{kz.json,Username})'!g"
 #!substdef "!KZQ_HANDLE_NEW_SUBSCRIBE_DELETE1!delete from active_watchers where callid = '\$ci'!g"
 #!substdef "!KZQ_HANDLE_NEW_SUBSCRIBE_DELETE2!delete from active_watchers where watcher_username='\$fU' and presentity_uri='\$var(presentity_uri)' and to_user='\$tU' and watcher_domain='\$fd' and event='\$hdr(Event)'!g"
+#!substdef "!KZQ_HAS_PRESENTITY!select count(*) as count from presentity where username = '\$subs(to_user)' and domain = '\$subs(to_domain)' and event = '\$subs(event)'!g"
```
Comment (lazedo) [inline kamailio/db_queries_postgres.cfg]: same here

File: kamailio/db_scripts/mysql-specific [hunk 5]
```diff
@@ -40,5 +40,6 @@ EOF
 sql_footer() {
 cat << EOF
 ALTER TABLE location_attrs MODIFY avalue varchar(512);
+ALTER TABLE presentity MODIFY etag varchar(128);
```
Comment (lazedo) [inline kamailio/db_scripts/mysql-specific]: why this change ?

File: kamailio/db_scripts/postgres-specific [hunk 6]
```diff
@@ -40,5 +40,6 @@ EOF
 sql_footer() {
 cat << EOF
 ALTER TABLE location_attrs ALTER COLUMN avalue TYPE varchar(512);
+ALTER TABLE presentity ALTER COLUMN etag TYPE varchar(128);
```
Comment (lazedo) [inline kamailio/db_scripts/postgres-specific]: maybe we should make theses changes  in kamailio ?

File: kamailio/db_queries_kazoo.cfg [hunk 7]
```diff
@@ -6,10 +6,11 @@
...
 #!substdef "!KZQ_PRESENCE_SEARCH_DETAIL!select * from active_watchers_log where presentity_uri = \"\$var(presentity_uri)\"!g"
 #!substdef "!KZQ_PRESENCE_SEARCH_SUMMARY!select * from active_watchers where watcher_domain = \"\$var(Domain)\"!g"
-#!substdef "!KZQ_HAS_PRESENTITY!select count(*) as count from presentity where username = \"\$subs(to_user)\" and domain = \"\$subs(to_domain)\" and event = \"\$subs(event)\"!g"
 #!substdef "!KZQ_PRESENCE_RESET!delete from presentity where sender = \"\$var(MediaUrl)\"!g"
+#!substdef "!KZQ_REPLACE_WATCHERS_LOG!REPLACE INTO active_watchers_log (presentity_uri, watcher_username, watcher_domain, event, callid, to_user, to_domain, user_agent, time, result, sent_msg, received_msg) VALUES (\"\$subs(uri)\", \"\$subs(watcher_username)\", \"\$subs(watcher_domain)\", \"\$subs(event)\",\"\$subs(callid)\",\"\$subs(to_user)\",\"\$subs(to_domain)\", '\$(subs(user_agent){s.escape.common}{s.replace,\\\',''}{s.replace,\$\$,})', \$TS, \$notify_reply(\$rs), '\$(mb{s.escape.common}{s.replace,\\\',''}{s.replace,\$\$,})', '\$(notify_reply(\$mb){s.escape.common}{s.replace,\\\',''}{s.replace,\$\$,})')|g"
```
Comment (lazedo) [inline kamailio/db_queries_kazoo.cfg]: this still doesn't work because of the ending `|g`

File: kamailio/db_queries_kazoo.cfg (hunk 2 above)
Comment (sergey-safarov) [reply kamailio/db_queries_kazoo.cfg]: I ordered macros for easy search by eyes

File: kamailio/db_queries_mysql.cfg (hunk 3 above)
Comment (sergey-safarov) [reply kamailio/db_queries_mysql.cfg]: I ordered macros for easy search by eyes

File: kamailio/db_queries_postgres.cfg (hunk 4 above)
Comment (sergey-safarov) [reply kamailio/db_queries_postgres.cfg]: I ordered macros for easy search by eyes

File: kamailio/db_scripts/mysql-specific (hunk 5 above)
Comment (sergey-safarov) [reply kamailio/db_scripts/mysql-specific]: When in tested i can see errors that value `etag` cannot be saved as too long value.
I adjusted presentity to fix this.

File: kamailio/db_scripts/postgres-specific (hunk 6 above)
Comment (sergey-safarov) [reply kamailio/db_scripts/postgres-specific]: Hello Daniel @miconda
Could you look this PR.
May be need to extend `etag` column in `presentity` table?
This is because i can see errors as current column size is shor to save `etag` value.
Optionally I can open ticket for this on `kamailio/kamailio` repo.

File: kamailio/db_queries_kazoo.cfg (hunk 7 above)
Comment (sergey-safarov) [reply kamailio/db_queries_kazoo.cfg]: Removed `|g`
Now macro defined as
```
#!substdef "!KZQ_REPLACE_WATCHERS_LOG!REPLACE INTO active_watchers_log (presentity_uri, watcher_username, watcher_domain, event, callid, to_user, to_domain, user_agent, time, result, sent_msg, received_msg) VALUES (\"\$subs(uri)\", \"\$subs(watcher_username)\", \"\$subs(watcher_domain)\", \"\$subs(event)\",\"\$subs(callid)\",\"\$subs(to_user)\",\"\$subs(to_domain)\", '\$(subs(user_agent){s.escape.common}{s.replace,\\\',''}{s.replace,\$\$,})', \$TS, \$notify_reply(\$rs), '\$(mb{s.escape.common}{s.replace,\\\',''}{s.replace,\$\$,})', '\$(notify_reply(\$mb){s.escape.common}{s.replace,\\\',''}{s.replace,\$\$,})')|"
```

File: kamailio/db_scripts/postgres-specific (hunk 6 above)
Comment (miconda) [reply kamailio/db_scripts/postgres-specific]: I just updated the schema definition in master so etag size is now 128 for presence tables.

File: kamailio/db_queries_kazoo.cfg [hunk 8]
```diff
@@ -6,10 +6,11 @@
...
 #!substdef "!KZQ_RESET_ACCOUNT_DELETE!delete from presentity where domain=\"\$(kzE{kz.json,Realm})\"!g"
 #!substdef "!KZQ_RESET_ACCOUNT_UPDATE!update active_watchers set expires = \$TS where watcher_domain=\"\$(kzE{kz.json,Realm})\"!g"
 #!substdef "!KZQ_RESET_PUBLISHER_UPDATE!update active_watchers set expires = \$TS where id in (select b.id from presentity a inner join active_watchers b on a.username = b.to_user and a.domain = b.to_domain and a.event = b.event where a.sender = \"\$var(MediaUrl)\")!g"
 #!substdef "!KZQ_PRESENCE_SEARCH_DETAIL!select * from active_watchers_log where presentity_uri = \"\$var(presentity_uri)\"!g"
 #!substdef "!KZQ_PRESENCE_SEARCH_SUMMARY!select * from active_watchers where watcher_domain = \"\$var(Domain)\"!g"
-#!substdef "!KZQ_HAS_PRESENTITY!select count(*) as count from presentity where username = \"\$subs(to_user)\" and domain = \"\$subs(to_domain)\" and event = \"\$subs(event)\"!g"
 #!substdef "!KZQ_PRESENCE_RESET!delete from presentity where sender = \"\$var(MediaUrl)\"!g"
```
Comment (sergey-safarov) [inline kamailio/db_queries_kazoo.cfg]: Looks as `KZQ_PRESENCE_RESET` also need to order

**Comment by lazedo** [comment]
@sergey-safarov this needs rebase, reopen when ready.
